gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -190,15 +195,24 @@ public TransportChannel getTransportChannel() throws IOException {
   private TransportChannel createChannel() throws IOException {
     ManagedChannel outerChannel;
 
-    if (poolSize == null || poolSize == 1) {
-      outerChannel = createSingleChannel();
+    int realPoolSize;
+    if (poolSize == null) {
+      realPoolSize = 1;
     } else {
-      ImmutableList.Builder<ManagedChannel> channels = ImmutableList.builder();
-
-      for (int i = 0; i < poolSize; i++) {
-        channels.add(createSingleChannel());
-      }
-      outerChannel = new ChannelPool(channels.build());
+      realPoolSize = poolSize;
+    }
+    ChannelFactory channelFactory =
+        new ChannelFactory() {
+          public ManagedChannel createSingleChannel() throws IOException {
+            return InstantiatingGrpcChannelProvider.this.createSingleChannel();
```

Comment:
This does not look pretty. This creates an essentially "fake" channel factory, which propagates calls to a private `createSingleChannel() ` method. Why not just make this class Implement `ChannelFactory`, and make that method public and then use this class (`InstantiatingGrpcChannelProvider`) as `ChannelFactory` (which it essentially is). Does making class implement another interface break binary compatibility?  I guess/hope no.

Reply:
I requested this approach: ChannelFactory is an internal api, while InstantiatingGrpcChannelProvider is public. Having the public api implement an internal and exposing it on the public surface is dirty. The anonymous class keeps the class internal